pkg/registry/core/service/storage/rest.go

```diff
@@ -199,7 +199,8 @@ func (rs *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 	nodePortOp := portallocator.StartOperation(rs.serviceNodePorts, dryrun.IsDryRun(options.DryRun))
 	defer nodePortOp.Finish()
 
-	if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer {
+	if service.Spec.Type == api.ServiceTypeNodePort ||
+		(service.Spec.Type == api.ServiceTypeLoadBalancer && (service.Spec.AllocateLoadBalancerNodePorts == nil || *service.Spec.AllocateLoadBalancerNodePorts)) {
```

Comment:
I think we also need to gate `updateNodePorts` on line 432 

https://github.com/kubernetes/kubernetes/blob/cd6c4921fc11d00dd60324341061a057da9cb54e/pkg/registry/core/service/storage/rest.go#L430-L434

Reply:
Yes, I have commented that myself. But I want to make sure that it is possible to remove the nodePort once the AllocateLoadBalancerNodePorts is set to "false" for a service.

Reply:
I might be remembering incorrectly, but I think in the KEP we said if a node port is already set and `AllocateLoadBalancerNodePorts=false`, then we should do nothing and the user needs to update the Service themselves to remove the allocated node port. This means `updateNodePorts` should not run when `AllocateLoadBalancerNodePorts=false`, this leaves the node port as-is but also allows user to remove it without the allocator re-allocating it on removal.

Reply:
You are absolutely right. I did not understand the update function fully.

Reply:
I will fix this and test it but first I must re-base/re-generate (sigh...)